targets/csharp/source/PlayFabSDK/source/Qos/PlayFabQosApi.cs

```diff
@@ -0,0 +1,163 @@
+﻿namespace PlayFab.QoS
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using EventsModels;
+    using Internal;
+    using MultiplayerModels;
+
+    public class PlayFabQosApi
+    {
+        private const int DefaultTimeoutMs = 250;
+        private readonly Dictionary<AzureRegion, string> _dataCenterMap = new Dictionary<AzureRegion, string>();
+        private readonly Action<string> _logAction;
+
+        public PlayFabQosApi(Action<string> logAction = null)
+        {
+            _logAction = logAction;
+        }
+
+        public async Task<QosResult> GetQosResultAsync(int timeoutMs = DefaultTimeoutMs)
+        {
+            QosResult result = await GetResultAsync(timeoutMs);
+            if (result.ErrorCode != (int)QosErrorCode.Success)
+            {
+                return result;
+            }
+
+            Task.Factory.StartNew(() => SendResultsToPlayFab(result));
+            return result;
+        }
+
+        private async Task<QosResult> GetResultAsync(int timeoutMs)
+        {
+            var result = new QosResult();
+
+            if (!PlayFabClientAPI.IsClientLoggedIn())
+            {
+                LogQos("Client is not logged in");
+                result.ErrorCode = (int)QosErrorCode.NotLoggedIn;
+                return result;
```

Comment:
I feel that you're making it unnecessarily harder for customers to determine what a QosResult actually means.
In this case, the information they need is scattered in multiple places. The int result means nothing unless they know it's actually a QosErrorCode. And the string is sent off to an entirely different place (logAction), almost always totally disconnected from this result.

You could add an ErrorMessage string field to QosResult, which is where "Client is not logged in" would fit perfectly, and it would be sent to the caller of this method, instead of somewhere else entirely. And changing QosResult.ErrorCode to QosErrorCode would reinforce the meaning of those values, and make it very easy for the caller to resolve the error.

Reply:
I created an ErrorMessage field. However I am still keeping the ErrorCode as int as pf-main is expecting an integer error code. 

---
In reply to: [288682801](https://github.com/PlayFab/SDKGenerator/pull/535#discussion_r288682801) [](ancestors = 288682801)